xdem/coreg.py

```diff
@@ -1664,55 +1715,154 @@ def _fit_func(self, ref_dem: np.ndarray, tba_dem: np.ndarray, transform: Optiona
 
-    def _apply_func(self, dem: np.ndarray, transform: rio.transform.Affine) -> np.ndarray:
-        """Apply the estimated x/y/z offsets to a DEM."""
-        bounds, resolution = _transform_to_bounds_and_res(dem.shape, transform)
-        scaling_factor = self._meta["resolution"] / resolution
+    def _to_matrix_func(self) -> np.ndarray:
+        """Return a transformation matrix from the estimated offsets."""
+        offset_east = self._meta["offset_east_px"] * self._meta["resolution"]
+        offset_north = self._meta["offset_north_px"] * self._meta["resolution"]
 
-        # Make index grids for the east and north dimensions
-        east_grid = np.arange(dem.shape[1]) * scaling_factor
-        north_grid = np.arange(dem.shape[0]) * scaling_factor
+        matrix = np.diag(np.ones(4, dtype=float))
+        matrix[0, 3] += offset_east
+        matrix[1, 3] += offset_north
+        matrix[2, 3] += self._meta["bias"]
 
-        # Make a function to estimate the DEM (used to construct an offset DEM)
-        elevation_function = scipy.interpolate.RectBivariateSpline(x=north_grid, y=east_grid,
-                                                                   z=np.where(np.isnan(dem), -9999, dem))
-        # Make a function to estimate nodata gaps in the aligned DEM (used to fix the estimated offset DEM)
-        nodata_function = scipy.interpolate.RectBivariateSpline(x=north_grid, y=east_grid, z=np.isnan(dem))
+        return matrix
 
-        shifted_east_grid = east_grid + self._meta["offset_east_px"]
-        shifted_north_grid = north_grid - self._meta["offset_north_px"]
 
-        shifted_dem = elevation_function(y=shifted_east_grid, x=shifted_north_grid)
-        new_nans = nodata_function(y=shifted_east_grid, x=shifted_north_grid)
-        shifted_dem[new_nans >= 1] = np.nan
+def invert_matrix(matrix: np.ndarray) -> np.ndarray:
+    """Invert a transformation matrix."""
+    with warnings.catch_warnings():
+        # Deprecation warning from pytransform3d. Let's hope that is fixed in the near future.
+        warnings.filterwarnings("ignore", message="`np.float` is a deprecated alias for the builtin `float`")
 
-        shifted_dem += self._meta["bias"]
+        # Invert the transform if wanted.
+        return pytransform3d.transformations.invert_transform(matrix)
 
-        return shifted_dem
 
-    def _apply_pts_func(self, coords: np.ndarray) -> np.ndarray:
-        """Apply the estimated x/y/z offsets to a set of points."""
-        offset_east = self._meta["offset_east_px"] * self._meta["resolution"]
-        offset_north = self._meta["offset_north_px"] * self._meta["resolution"]
+def apply_matrix(dem: np.ndarray, transform: rio.transform.Affine, matrix: np.ndarray, invert: bool = False,
+                 centroid: Optional[tuple[float, float, float]] = None,
+                 resampling: Union[int, str] = "cubic") -> np.ndarray:
+    """
+    Apply a transformation matrix to a DEM.
```

Comment:
Can you maybe explain here the different steps to get the result? It is definitely non trivial, and probably not very standard.

Reply:
I tried to explain in 0c9d568. Please let me know if you have any suggestions for improvement!